Add function for update x to ideal x.

seigodev opened the pull request:
座標決定機能の追加 #15
update_x2idealx(): ノードのx座標の更新を行う。更新には優先度法を用いた。
assign_x_continuously(): スタック内のノードに順にx座標を割り当てる。

seigodev committed:
Add function for update x to ideal x.

座標決定機能の追加 #15
update_x2idealx():ノードの座標の更新を行う。
assign_x_continuously(): スタック内のノードに順にx座標を割り当てる。

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+            assign_x_continuously(node_stack, ideal_x, not larger_idealx)
+
+
+def assign_x_continuously(nodes_stack, x, increment):
```
assign_x_continously -> assign_x_in_sequence
increment -> sign (+1 or -1)

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+        if increment:
+            x += 1
+        else:
+            x -= 1
```
x += sign

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
         node2idealx_dict[node] = idealx
 
 
+def update_x2idealx(node, same_level_nodes, ideal_x, assigned_nodes, larger_idealx):
```
larger_idealx -> sign (+1 or -1)

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+    node_stack = Stack()
+    node_stack.push(node)
+    while node_stack.is_empty() is False:
+        update_x = node.x + 1 if larger_idealx else node.x - 1
```
```python
update_x = node.x + sign
```

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+            if other.x == update_x:  # 更新先にノードがあるかを調べる
+                exist = True
+                if other in assigned_nodes:
+                    assign_x_continuously(node_stack, node.x, not larger_idealx)
```
```python
assign_x_in_sequence(node_stack, node.x, -sign)
```

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+    while node_stack.is_empty() is False:
+        update_x = node.x + 1 if larger_idealx else node.x - 1
+        exist = False  # 更新先にノードがあるかを見るフラグ
+        for other in same_level_nodes:
```
このfor文は遅くないか？

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+                else:
+                    node_stack.push(other)
+                    node = other
+                    ideal_x = ideal_x + 1 if larger_idealx else ideal_x - 1
```
ideal_x += sign

aabaa commented on demo/create_graph.py:
```diff
@@ -578,6 +578,57 @@ def update_idealx(node2idealx_dict):
+        if exist is False:
+            node.x = update_x
+        if node.x == ideal_x:
+            assign_x_continuously(node_stack, ideal_x, not larger_idealx)
```
assign_x_in_sequence(node_stack, node.x, -sign)

seigodev committed:
Fix requested changes.

・increment -> sign（581, 628行）
・ssign_x_continuously -> assign_x_in_sequence （616行）
・ update_x2idealxのfor文について：
　再帰関数にして、while分を無くし、二重ループを無くしました。
　594-597行の処理は、端のノードをノードが存在しないほうへ更新しようとした際の処理です。
　for文は「if same_level_nodes[node_index+sign].x == update_x:」を使用することで削除されました。

aabaa commented on demo/create_graph.py:
```diff
@@ -578,55 +578,54 @@ def update_idealx(node2idealx_dict):
+        node_index: x座標を更新したいノードのsame_level_nodesにおけるインデックス
         same_level_nodes: nodeと同じ階層のノード
         ideal_x: nodeの理想のx座標値
         assigned_nodes: 既に割り当てを行ったノードのスタック
```
既に割当を行ったので動かしたくないノードが入ったスタック

aabaa commented on demo/create_graph.py:
```diff
@@ -578,55 +578,54 @@ def update_idealx(node2idealx_dict):
         ideal_x: nodeの理想のx座標値
         assigned_nodes: 既に割り当てを行ったノードのスタック
-        larger_idealx: 理想x座標が今のx座標より大きいか否か
+        node_stack: 座標を更新するノードの入ったスタック。
```
座標を更新している途中のノードが入ったスタック

aabaa commented on demo/create_graph.py:
```diff
@@ -578,55 +578,54 @@ def update_idealx(node2idealx_dict):
-
-
-def assign_x_continuously(nodes_stack, x, increment):
+    # オーバーフローする時の処理
```
ノード列の端に到達したため，走査を終了してnode_stackに入ったノードを理想x座標まで動かす

aabaa commented on demo/create_graph.py:
```diff
@@ -578,55 +578,54 @@ def update_idealx(node2idealx_dict):
+            node_index += sign
+            ideal_x += sign
+    else:
+        same_level_nodes[node_index].x = update_x
```
最後にまとめて割り当てるように変更する

seigodev committed:
Fix update_x2idealx_recursively().

・説明、アルゴリズムの追加。
・各引数の説明の変更
・update_x -> next_x
・更新ノードのx座標と理想x座標値の比較の処理を最初に(601-603行)

seigodev committed:
Rewrote update_x2idealx_recursively().

更新したいノードの隣のx座標を見ずに、隣のノードのx座標を見て、座標を更新するようにしました。
